src/main/java/net/minecraftforge/common/attachment/AttachmentType.java

```diff
@@ -0,0 +1,128 @@
+package net.minecraftforge.common.attachment;
+
+import com.mojang.serialization.Codec;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.nbt.Tag;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.common.util.INBTSerializable;
+import net.minecraftforge.registries.ForgeRegistries;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * Represents a data attachment type: some data that can be added to any object implementing {@link IAttachmentHolder}.
+ * <p>
+ * Data attachment types must be registered to {@link ForgeRegistries.Keys#ATTACHMENT_TYPES the registry}.
+ * <p>
+ * <h3>{@link ItemStack}-exclusive behavior:</h3>
+ * <ul>
+ *     <li>Serializable item stack attachments are synced between the server and the client.</li>
+ *     <li>Serializable item stack attachments are copied when an item stack is copied.</li>
+ *     <li>Serializable item stack attachments must match for item stack comparison to succeed.</li>
+ * </ul>
+ */
+// TODO: copy strategies for entities, to optionally persist attachments when restoring an entity (e.g. after player death)
+// TODO: maybe add copy handlers for stacks and entities, to customize copying behavior (instead of serializing, copying the NBT, deserializing)
```

Comment:
I would be interested in this, too - for `ItemStack`s mostly - as they tend to get copied a fair amount.

Reply:
Also future work IMO, as that can be added nicely a bit later!